tdm/docs/architecture.md

```diff
@@ -0,0 +1,222 @@
+
+
+## Goals
+
+In order to achieve high performance Input/Output (I/O) operations, confidential
+VMs must extend their trust boundary to include a composition of directly
+assigned, TEE-IO-capable devices/functions. Without this ability, confidential
+VMs have to resort to using para-virtualized I/O using non-confidential memory
+regions, which has performance impacts due to memory copies and negates the use
+of directly assigned devices.
+
+Silicon vendors and device manufacturers provide TEE-IO compliant hardware
+implementations. Intel TDX Connect or AMD SEV-TIO, for example, rely on device
+manufacturers to support the TDISP, SPDM and PCIe IDE protocols to give
+confidential guests access to trusted MMIO and secure DMA between them and
+devices which trustworthiness is verifiable.
+
+The TEE-IO security model is built on top of this hardware support but also on
+the ability for confidential guests to ultimately decide if they accept or
+reject TEE-IO compliant devices into their TCB. The TDM, as a confidential guest
+component, is the TCB acceptance decision maker and is responsible for notifying
+the underlying Trusted Security Manager (TSM) about its decision. By accepting a
+TDI into the guest TCB, the TDM explicitly requests the TSM to move the device
+into an operational state. The TDM builds the acceptance decision after having
+verified and authenticated the TDI. It must also attest of the assigned device
+trustworthiness, through a local or remote attestation process.
+
+## Scope
+
+## Architectural Overview
+
+High-level view:
+
+```
+
+
+                                   ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+                                   ┃                            Confidential Guest ┃
+                                   ┃           ╔═══════════════════════╗           ┃
+                                   ┃           ║Trusted Device Manager ║           ┃
+                                   ┃           ║                       ║           ┃
+                                   ┃           ║┌─────────────────────┐║           ┃
+                                   ┃           ║│    Relying Party    │║           ┃
+                                   ┃      ┌───▷║│(Attestation Service)│║           ┃
+                                   ┃      │    ║└─────────────────────┘║           ┃
+                                   ┃      │    ║                       ║           ┃
+                                   ┃      │    ║ ╔════════╗ ╔════════╗ ║           ┃
+                                   ┃      │    ╚═╣  TEE   ╠═╣  TEE   ╠═╝           ┃
+                                   ┃      │      ║ Plugin ║ ║ Plugin ║             ┃
+                                   ┃     udev    ╚════════╝ ╚════════╝             ┃
```

Comment:
So the TDM will be modular at the level of the TEE? What about different devices? Will they all be attested with the same code?

Reply:
Yes, I see the TEE interface as the ABI beween the guest and the TSM. Today we use it mostly for fetching attestation evidences, but I think it should/will be extended to also support the TEE-IO flows.
All assigned devices will be managed by the TSM, on behalf of the untrusted VMM, so the TSM will be the single interface for getting any device attestation evidence, and for guests to accept devices into their TCBs.
Does that make sense?

Reply:
@sameo @fitzthum I think a better approach may be to unify the TEE plugin here with the existing attester plugin in AA, both of which are at the same level as the `attester` clause module. Here, we can import the `attester` crate to obtain evidence or call AA's `get_evidence` API to obtain evidence.